src/datomish/clauses.cljc

```diff
@@ -0,0 +1,256 @@
+  (when-not (instance? DefaultSrc (:source pattern))
+    (raise-str "Non-default sources are not supported in patterns. Pattern: " pattern))
+
+  (let [[table alias] (source->from (:source cc))   ; e.g., [:datoms :datoms123]
+        places (map (fn [place col] [place col])
+                    (:pattern pattern)
+                    (:columns (:source cc)))]
+    (reduce
+      (fn [cc
+           [pattern-part                           ; ?x, :foo/bar, 42
+            position]]                             ; :a
+        (let [col (sql/qualify alias (name position))]    ; :datoms123.a
+          (condp instance? pattern-part
+            ;; Placeholders don't contribute any bindings, nor do
+            ;; they constrain the query -- there's no need to produce
+            ;; IS NOT NULL, because we don't store nulls in our schema.
+            Placeholder
+            cc
+
+            Variable
+            (bind-column-to-var cc pattern-part col)
+
+            Constant
+            (constrain-column-to-constant cc col position (:value pattern-part))
+
+            (raise-str "Unknown pattern part " pattern-part))))
+
+      ;; Record the new table mapping.
+      (util/conj-in cc [:from] [table alias])
+
+      places)))
+
+(defn- plain-symbol->sql-predicate-symbol [fn]
+  (when-not (instance? PlainSymbol fn)
+    (raise-str "Predicate functions must be named by plain symbols." fn))
+  (#{:> :< :=} (keyword (name (:symbol fn)))))
+
+(defn apply-predicate-clause [cc predicate]
+  (when-not (instance? Predicate predicate)
+    (raise-str "Expected to be called with a Predicate instance." predicate))
+  (let [f (plain-symbol->sql-predicate-symbol (:fn predicate))]
+    (when-not f
+      (raise-str "Unknown function " (:fn predicate)))
+
+    (let [args (map
+                 (fn [arg]
+                   (condp instance? arg
+                     Placeholder
+                     (raise-str "Can't use a placeholder in a predicate.")
+
+                     Variable
+                     (let [v (:symbol arg)
+                           internal-bindings (v (:bindings cc))
+                           external-bindings (v (:external-bindings cc))]
+                       (or (first internal-bindings)
+                           (first external-bindings)
+                           (raise-str "No bindings yet for " v)))
+
+                     Constant
+                     (constant-in-source (:source cc) (:value arg))
+
+                     (raise-str "Unknown predicate argument " arg)))
+
+                 (:args predicate))]
+      (util/conj-in cc [:wheres] (cons f args)))))
+
+(defn apply-not-clause [cc not]
+  (when-not (instance? Not not)
+    (raise-str "Expected to be called with a Not instance." not))
+  (when-not (instance? DefaultSrc (:source not))
+    (raise-str "Non-default sources are not supported in patterns. Pattern: " not))
+
+  ;; If our bindings are already available, great -- emit a :wheres
```

Comment:
DS and Datomic do not re-order.  No need for us to be fancy: straight left-to-right.